### PR TITLE
LibJS: New bytecode format, remove JIT compiler

### PR DESCRIPTION
**Pull request by awesomekling**
The JIT compiler was an interesting experiment, but ultimately the
security & complexity cost of doing arbitrary code generation at runtime
is far too high.

Other engines, JavaScriptCore in particular, have already proven that
it's possible to handle the vast majority of contemporary web content
with an interpreter. They are currently ~5x faster than us on benchmarks
when running without a JIT. We need to catch up to them before
considering performance techniques with a heavy security cost.

**LibJS/Bytecode: Move to a new bytecode format**
    
This patch moves us away from the accumulator-based bytecode format to
one with explicit source and destination registers.
    
The new format has multiple benefits:
    
- ~25% faster on the Kraken and Octane benchmarks :^)
- Fewer instructions to accomplish the same thing
- Much easier for humans to read(!)
    
Because this change requires a fundamental shift in how bytecode is
generated, it is quite comprehensive.
    
Main implementation mechanism: generate_bytecode() virtual function now
takes an optional "preferred dst" operand, which allows callers to
communicate when they have an operand that would be optimal for the
result to go into. It also returns an optional "actual dst" operand,
which is where the completion value (if any) of the AST node is stored
after the node has "executed".
    
One thing of note that's new: because instructions can now take locals
as operands, this means we got rid of the GetLocal instruction.
A side-effect of that is we have to think about the temporal deadzone
(TDZ) a bit differently for locals (GetLocal would previously check
for empty values and interpret that as a TDZ access and throw).
We now insert special ThrowIfTDZ instructions in places where a local
access may be in the TDZ, to maintain the correct behavior.
    
There are a number of progressions and regressions from this test:
    
A number of async generator tests have been accidentally fixed while
converting the implementation to the new bytecode format. It didn't
seem useful to preserve bugs in the original code when converting it.
    
Some "does eval() return the correct completion value" tests have
regressed, in particular ones related to propagating the appropriate
completion after control flow statements like continue and break.
These are all fairly obscure issues, and I believe we can continue
working on them separately.
   
The net test262 result is a progression though. :^)

```
Summary:
    Diff Tests:
        +6 ✅    -6 ❌   

Diff Tests:
    test/built-ins/AsyncFromSyncIteratorPrototype/return/iterator-result-unwrap-promise.js              ❌ -> ✅
    test/built-ins/AsyncFromSyncIteratorPrototype/return/iterator-result.js                             ❌ -> ✅
    test/built-ins/AsyncFromSyncIteratorPrototype/return/return-undefined.js                            ❌ -> ✅
    test/language/expressions/async-generator/named-yield-star-async-return.js                          ❌ -> ✅
    test/language/expressions/async-generator/named-yield-star-sync-return.js                           ❌ -> ✅
    test/language/expressions/async-generator/yield-star-async-return.js                                ❌ -> ✅
    test/language/expressions/async-generator/yield-star-sync-return.js                                 ❌ -> ✅
    test/language/expressions/class/async-gen-method-static/yield-star-async-return.js                  ❌ -> ✅
    test/language/expressions/class/async-gen-method-static/yield-star-sync-return.js                   ❌ -> ✅
    test/language/expressions/class/async-gen-method/yield-star-async-return.js                         ❌ -> ✅
    test/language/expressions/class/async-gen-method/yield-star-sync-return.js                          ❌ -> ✅
    test/language/expressions/class/elements/async-gen-private-method-static/yield-star-async-return.js ❌ -> ✅
    test/language/expressions/class/elements/async-gen-private-method-static/yield-star-sync-return.js  ❌ -> ✅
    test/language/expressions/class/elements/async-gen-private-method/yield-star-async-return.js        ❌ -> ✅
    test/language/expressions/class/elements/async-gen-private-method/yield-star-sync-return.js         ❌ -> ✅
    test/language/expressions/object/method-definition/async-gen-yield-star-async-return.js             ❌ -> ✅
    test/language/expressions/object/method-definition/async-gen-yield-star-sync-return.js              ❌ -> ✅
    test/language/expressions/yield/star-rhs-iter-rtrn-res-value-err.js                                 ❌ -> ✅
    test/language/statements/async-generator/yield-star-async-return.js                                 ❌ -> ✅
    test/language/statements/async-generator/yield-star-sync-return.js                                  ❌ -> ✅
    test/language/statements/class/async-gen-method-static/yield-star-async-return.js                   ❌ -> ✅
    test/language/statements/class/async-gen-method-static/yield-star-sync-return.js                    ❌ -> ✅
    test/language/statements/class/async-gen-method/yield-star-async-return.js                          ❌ -> ✅
    test/language/statements/class/async-gen-method/yield-star-sync-return.js                           ❌ -> ✅
    test/language/statements/class/elements/async-gen-private-method-static/yield-star-async-return.js  ❌ -> ✅
    test/language/statements/class/elements/async-gen-private-method-static/yield-star-sync-return.js   ❌ -> ✅
    test/language/statements/class/elements/async-gen-private-method/yield-star-async-return.js         ❌ -> ✅
    test/language/statements/class/elements/async-gen-private-method/yield-star-sync-return.js          ❌ -> ✅
    test/language/statements/do-while/S12.6.1_A5.js                                                     ✅ -> ❌
    test/language/statements/do-while/cptn-abrupt-empty.js                                              ✅ -> ❌
    test/language/statements/for/cptn-decl-expr-iter.js                                                 ✅ -> ❌
    test/language/statements/for/cptn-decl-expr-no-iter.js                                              ✅ -> ❌
    test/language/statements/for/cptn-expr-expr-iter.js                                                 ✅ -> ❌
    test/language/statements/for/cptn-expr-expr-no-iter.js                                              ✅ -> ❌
    test/language/statements/if/cptn-no-else-true-abrupt-empty.js                                       ✅ -> ❌
    test/language/statements/switch/cptn-a-abrupt-empty.js                                              ✅ -> ❌
    test/language/statements/switch/cptn-a-fall-thru-abrupt-empty.js                                    ✅ -> ❌
    test/language/statements/switch/cptn-b-abrupt-empty.js                                              ✅ -> ❌
    test/language/statements/switch/cptn-b-fall-thru-abrupt-empty.js                                    ✅ -> ❌
    test/language/statements/switch/cptn-dflt-abrupt-empty.js                                           ✅ -> ❌
    test/language/statements/switch/cptn-dflt-b-abrupt-empty.js                                         ✅ -> ❌
    test/language/statements/switch/cptn-dflt-b-fall-thru-abrupt-empty.js                               ✅ -> ❌
    test/language/statements/switch/cptn-dflt-fall-thru-abrupt-empty.js                                 ✅ -> ❌
    test/language/statements/switch/cptn-no-dflt-match-abrupt-empty.js                                  ✅ -> ❌
    test/language/statements/switch/cptn-no-dflt-match-fall-thru-abrupt-empty.js                        ✅ -> ❌
    test/language/statements/try/cptn-catch-empty-break.js                                              ✅ -> ❌
    test/language/statements/try/cptn-catch-empty-continue.js                                           ✅ -> ❌
    test/language/statements/try/cptn-catch-finally-empty-break.js                                      ✅ -> ❌
    test/language/statements/try/cptn-catch-finally-empty-continue.js                                   ✅ -> ❌
    test/language/statements/try/cptn-finally-empty-break.js                                            ✅ -> ❌
    test/language/statements/try/cptn-finally-empty-continue.js                                         ✅ -> ❌
    test/language/statements/try/cptn-finally-from-catch.js                                             ❌ -> ✅
    test/language/statements/try/cptn-finally-skip-catch.js                                             ❌ -> ✅
    test/language/statements/try/cptn-finally-wo-catch.js                                               ❌ -> ✅
    test/language/statements/try/cptn-try.js                                                            ❌ -> ✅
    test/language/statements/while/S12.6.2_A5.js                                                        ✅ -> ❌
    test/language/statements/while/cptn-abrupt-empty.js                                                 ✅ -> ❌
    test/language/statements/with/cptn-abrupt-empty.js                                                  ✅ -> ❌
```

Benchmark results (old/new bytecode):
```
Suite       Test                                   Speedup  Old (Mean ± Range)           New (Mean ± Range)
----------  -----------------------------------  ---------  ---------------------------  ---------------------------
Kraken      ai-astar.js                              1.197  5.250 ± 5.190 … 5.310        4.385 ± 4.380 … 4.390
Kraken      audio-beat-detection.js                  1.45   3.495 ± 3.460 … 3.530        2.410 ± 2.410 … 2.410
Kraken      audio-dft.js                             1.17   2.585 ± 2.570 … 2.600        2.210 ± 2.200 … 2.220
Kraken      audio-fft.js                             1.478  3.400 ± 3.380 … 3.420        2.300 ± 2.300 … 2.300
Kraken      audio-oscillator.js                      1.108  2.675 ± 2.670 … 2.680        2.415 ± 2.410 … 2.420
Kraken      imaging-darkroom.js                      1.213  4.810 ± 4.800 … 4.820        3.965 ± 3.910 … 4.020
Kraken      imaging-desaturate.js                    1.559  5.380 ± 5.300 … 5.460        3.450 ± 3.440 … 3.460
Kraken      imaging-gaussian-blur.js                 1.26   18.075 ± 17.490 … 18.660     14.345 ± 14.330 … 14.360
Kraken      json-parse-financial.js                  0.962  0.125 ± 0.120 … 0.130        0.130 ± 0.130 … 0.130
Kraken      json-stringify-tinderbox.js              0.882  0.150 ± 0.150 … 0.150        0.170 ± 0.170 … 0.170
Kraken      stanford-crypto-aes.js                   1.363  1.540 ± 1.540 … 1.540        1.130 ± 1.130 … 1.130
Kraken      stanford-crypto-ccm.js                   1.268  1.160 ± 1.160 … 1.160        0.915 ± 0.910 … 0.920
Kraken      stanford-crypto-pbkdf2.js                1.344  2.970 ± 2.970 … 2.970        2.210 ± 2.190 … 2.230
Kraken      stanford-crypto-sha256-iterative.js      1.258  1.000 ± 0.990 … 1.010        0.795 ± 0.790 … 0.800
Octane      box2d.js                                 1.019  4.230 ± 4.210 … 4.250        4.150 ± 4.100 … 4.200
Octane      code-load.js                             1.005  2.125 ± 2.120 … 2.130        2.115 ± 2.110 … 2.120
Octane      crypto.js                                1.213  12.665 ± 12.660 … 12.670     10.440 ± 10.400 … 10.480
Octane      deltablue.js                             0.993  3.045 ± 3.030 … 3.060        3.065 ± 3.040 … 3.090
Octane      earley-boyer.js                          0.98   23.015 ± 22.980 … 23.050     23.475 ± 23.380 … 23.570
Octane      gbemu.js                                 1.117  6.195 ± 6.140 … 6.250        5.545 ± 5.490 … 5.600
Octane      mandreel.js                              1.198  35.480 ± 35.310 … 35.650     29.620 ± 29.550 … 29.690
Octane      navier-stokes.js                         2.521  5.760 ± 5.700 … 5.820        2.285 ± 2.280 … 2.290
Octane      pdfjs.js                                 1.041  3.920 ± 3.910 … 3.930        3.765 ± 3.760 … 3.770
Octane      raytrace.js                              0.998  8.205 ± 8.190 … 8.220        8.225 ± 8.150 … 8.300
Octane      regexp.js                                0.976  24.230 ± 23.970 … 24.490     24.835 ± 24.780 … 24.890
Octane      richards.js                              1      2.025 ± 2.020 … 2.030        2.025 ± 2.020 … 2.030
Octane      splay.js                                 0.997  2.980 ± 2.970 … 2.990        2.990 ± 2.990 … 2.990
Octane      typescript.js                            0.992  52.580 ± 52.560 … 52.600     53.010 ± 52.520 … 53.500
Octane      zlib.js                                  1.416  305.330 ± 304.600 … 306.060  215.570 ± 213.420 … 217.720
Kraken      Total                                    1.289  52.615                       40.830
Octane      Total                                    1.257  491.785                      391.115
All Suites  Total                                    1.26   544.400                      431.945
```

### CODE REVIEW

File: Userland/Libraries/LibJS/Bytecode/Generator.h [hunk 1]
```diff
@@ -241,6 +241,16 @@ class Generator {
     [[nodiscard]] size_t next_environment_variable_cache() { return m_next_environment_variable_cache++; }
     [[nodiscard]] size_t next_property_lookup_cache() { return m_next_property_lookup_cache++; }
 
+    [[nodiscard]] Operand add_constant(Value value)
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Bytecode/Generator.h]: Ignore if we're not worried about atomicity for this PR, but `Operand` does not exist as of this commit

File: Userland/Libraries/LibJS/Bytecode/Operand.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    Operand() = default;
+    [[nodiscard]] bool operator==(Operand const&) const = default;
+
+    explicit Operand(Type type, u32 index)
+        : m_type(type)
+        , m_index(index)
+    {
+    }
+
+    explicit Operand(Register);
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Bytecode/Operand.h:30]: Does this ctor have an implementation?

File: Userland/Libraries/LibJS/Bytecode/Operand.h [hunk 3]
```diff
@@ -14,13 +14,11 @@ namespace JS::Bytecode {
 class Operand {
 public:
     enum class Type {
-        Invalid,
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Bytecode/Operand.h]: Changes in this file in this commit seem like they should be squashed into the "add Operand" commit

File: Userland/Libraries/LibJS/Bytecode/Operand.h (hunk 2 above)
Comment (trflynn89) [reply Userland/Libraries/LibJS/Bytecode/Operand.h:30]: ah, that ctor is implemented a few commits later

**Review by trflynn89 — COMMENTED** [review]
The last commit is a lot to chew but seems good to me! At least I don't seem anything glaringly out of sorts.

A couple things that stuck out to me:
* The `Optional<Bytecode::Operand>` parameters seem like they should be passed const-ref. clangd seems to suggest as much, although it always seemed weird to me that small `Optional<T>`s shouldn't be passed by value. Maybe due to some triviality trait on Optional itself.

* We do this kinda wordy operation a lot:
```c++
auto object = preferred_dst.has_value() ? preferred_dst.value() : Bytecode::Operand(generator.allocate_register());
```
Might be nicer to make an inlined helper for that, but feel free to ignore.

**Comment by skyrising** [comment]
Love the new bytecode format, but as someone who poured a bit of work into the JIT it's sad to see it go.

While a 1.25x speedup is nice, the JIT did provide a ~2.4x speedup (5.4x Kraken, 2.3x Octane), so I can't help but wonder what these the combination of new bytecode and JIT could achieve.

Wouldn't it be nicer to just disable the JIT for now, allowing somebody to fix it up, instead of outright removing it? (We could still remove it later if nobody wants to put in the work).

Obviously removing it is way easier than keeping it working at the moment, but that also makes it harder for anyone wanting to bring it back in the future.

**Comment by awesomekling** [comment]
> Wouldn't it be nicer to only disable the JIT for now, allowing somebody to fix it up, instead of outright removing it? (We could still remove it later if nobody wants to put in the work).

The JIT is still there in git history for anyone who wants to do something with it. :)

File: Userland/Libraries/LibJS/Bytecode/Interpreter.cpp [hunk 4]
```diff
@@ -77,10 +115,10 @@ Value Interpreter::get(Operand op) const
     case Operand::Type::Constant:
         return current_executable().constants[op.index()];
     }
-    VERIFY_NOT_REACHED();
+    __builtin_unreachable();
```
[the remark this answers is not in the document]
Comment (awesomekling) [reply Userland/Libraries/LibJS/Bytecode/Interpreter.cpp:118]: :thinking: perhaps..

**Comment by awesomekling** [comment]
> The last commit is a lot to chew but seems good to me! At least I don't seem anything glaringly out of sorts.
> 
> A couple things that stuck out to me: ...

Took care of all your comments, I think!

**Review by trflynn89 — APPROVED** [review]
:star:

**Pull request merged**